bookwyrm/views/inbox.py

```diff
@@ -130,15 +130,16 @@ def has_valid_signature(request, activity):
     """verify incoming signature"""
     try:
         signature = Signature.parse(request)
-
-        key_actor = urldefrag(signature.key_id).url
-        if key_actor != activity.get("actor"):
-            raise ValueError("Wrong actor created signature.")
-
-        remote_user = activitypub.resolve_remote_id(key_actor, model=models.User)
+        remote_user = activitypub.resolve_remote_id(
+            activity.get("actor"), model=models.User
+        )
         if not remote_user:
             return False
 
+        if signature.key_id != remote_user.key_pair.remote_id:
+            if signature.key_id != f"{remote_user.remote_id}#main-key": # legacy Bookwyrm
```

Comment:
I'm a bit confused by this, isn't the legacy one the one with the `/`? And I thought `remote_user.remote_id` didn't have a `/`?

Reply:
It is incredibly confusing.

All Bookwyrm keyIds are at `{username}/#main-key` - this is set in `models.User` so it's impractical to change that because we can't retrospectively change people's keyIds.

However, the legacy _signatures_ (falsely) assert that the keyId is at `{username}#main-key`. Both are equivalent if you want to hit a url, but are not if you want to compare strings.

So if the remote user is from another Bookwyrm instance that has not upgraded, it will send a signature asserting the remote user's keyId is at `#main-key` but when our updated instance goes and checks, it will see that the keyId is actually listed as `/#main-key`. This line is a bit of a hack where our instance says "hmm, maybe it's old Bookwyrm, what if I just pretend the last forward slash isn't there?".